File: 853.car-fleet.py
#
# @lc app=leetcode id=853 lang=python3
#
# [853] Car Fleet
#

# @lc code=start
from collections import deque


class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        pairs = sorted(zip(position, speed), reverse=True) # furthest is bottleneck
        p1, s1 = pairs[0]
        stack = deque([(p1, (target - p1) / s1)])

        for p, s in pairs[1:]:
            t = (target - p) / s # time to reach end
            if t > stack[-1][1]: # can't catch up, new bottleneck / fleet
                stack.append((p, t))
        return len(stack)


# @lc code=end

File: RustSolutions/src/853.car-fleet.rs
/*
 * @lc app=leetcode id=853 lang=rust
 *
 * [853] Car Fleet
 */

// @lc code=start

use std::collections::VecDeque;

impl Solution {
    pub fn car_fleet(target: i32, position: Vec<i32>, speed: Vec<i32>) -> i32 {
        let mut pairs = position
            .into_iter()
            .zip(speed.into_iter())
            .collect::<Vec<_>>();
        pairs.sort_by(|(p1, _), (p2, _)| p2.cmp(p1));

        let (p1, s1) = pairs[0];
        let mut stack = VecDeque::new();
        stack.push_back((p1, (target - p1) as f64 / s1 as f64));

        for &(p, s) in &pairs[1..] {
            let t = (target - p) as f64 / s as f64;
            if t > stack.back().unwrap().1 {
                stack.push_back((p, t));
            }
        }
        stack.len() as i32
    }
}
// @lc code=end
